x/airdrop/keeper/claim.go

```diff
@@ -0,0 +1,195 @@
+	todaysIndex, err := airdrop.GetCurrentDateIndex(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Sum the rewards up to that date and 0 them out in the process
+	todaysRewards := sdkmath.ZeroInt()
+	for i := 0; i <= todaysIndex; i++ {
+		rewardsOnDate := userAllocation.Allocations[i]
+		todaysRewards = todaysRewards.Add(rewardsOnDate)
+		userAllocation.Allocations[i] = sdkmath.ZeroInt()
+	}
+
+	// If there are no rewards, alert the user with an error
+	if todaysRewards.IsZero() {
+		return types.ErrNoUnclaimedRewards
+	}
+
+	// Update the amount claimed on the allocation record
+	userAllocation.Claimed = userAllocation.Claimed.Add(todaysRewards)
+
+	// Distribute rewards from the distributor
+	distributorAccount := sdk.MustAccAddressFromBech32(airdrop.DistributionAddress)
+	claimerAccount := sdk.MustAccAddressFromBech32(userAllocation.Address)
+	rewardsCoin := sdk.NewCoin(airdrop.RewardDenom, todaysRewards)
+
+	if err := k.bankKeeper.SendCoins(ctx, distributorAccount, claimerAccount, sdk.NewCoins(rewardsCoin)); err != nil {
+		return errorsmod.Wrapf(err, "unable to distribute rewards")
+	}
+
+	// Update the reward record for to mark the progress
+	k.SetUserAllocation(ctx, userAllocation)
+
+	return nil
+}
+
+// User transaction to claim a portion of their total amount now, and forfeit the
+// remainder to be clawed back
+func (k Keeper) ClaimEarly(ctx sdk.Context, airdropId, claimer string) error {
+	// Fetch the airdrop and user's allocations
+	airdrop, airdropFound := k.GetAirdrop(ctx, airdropId)
+	if !airdropFound {
+		return types.ErrAirdropNotFound.Wrapf("airdrop %s", airdropId)
+	}
+	userAllocation, userFound := k.GetUserAllocation(ctx, airdropId, claimer)
+	if !userFound {
+		return types.ErrUserAllocationNotFound.Wrapf("user %s for airdrop %s", claimer, airdropId)
+	}
+
+	// Confirm the user has not elected the non-daily claim types
+	if userAllocation.ClaimType != types.CLAIM_DAILY {
+		return types.ErrClaimTypeUnavailable.Wrapf("user has already elected claim option %s",
+			userAllocation.ClaimType.String())
+	}
+
+	// Confirm we're not past the decision date
+	currentTime := ctx.BlockTime().Unix()
+	if currentTime > airdrop.ClaimTypeDeadlineDate.Unix() {
+		return types.ErrAfterDecisionDeadline
+	}
+
+	// Confirm the airdrop started
+	if currentTime < airdrop.DistributionStartDate.Unix() {
+		return types.ErrDistributionNotStarted
+	}
+
+	// Sum the total rewards 0 them out in the process
+	totalAccruedRewards := sdkmath.ZeroInt()
+	for i, rewardsOnDate := range userAllocation.Allocations {
+		totalAccruedRewards = totalAccruedRewards.Add(rewardsOnDate)
+		userAllocation.Allocations[i] = sdkmath.ZeroInt()
+	}
+
+	// If there are no rewards, alert the user with an error
+	if totalAccruedRewards.IsZero() {
+		return types.ErrNoUnclaimedRewards
+	}
+
+	// Calculate rewards after claim early penalty
+	rewardsRemainingRate := sdk.OneDec().Sub(airdrop.EarlyClaimPenalty)
```

Comment:
nit: should we sanity check this is /in [0,1]? ignore if we already have such checks on EarlyClaimPenalty

Reply:
it's handled in validate basic!